tests/docs/test_supported_confs.py

```diff
@@ -0,0 +1,493 @@
+    @pytest.mark.parametrize("shots", shotss)
+    def test_all_adjoint_nonexp(self, interface, return_type, shots):
+        """Test diff_method=adjoint raises an error for non-expectation
+        measurements for all interfaces"""
+        msg = "Adjoint differentiation method does not support measurement .*"
+
+        warn_msg = (
+            "Requested adjoint differentiation to be computed with finite shots. "
+            "Adjoint differentiation always calculated exactly."
+        )
+
+        with pytest.raises(QuantumFunctionError, match=msg):
+            with pytest.warns(UserWarning, match=warn_msg):
+                circuit = get_qnode(interface, "adjoint", return_type, shots)
+                x = get_variable(interface)
+                grad = compute_gradient(x, interface, circuit, return_type)
+
+    @pytest.mark.parametrize("interface", diff_interfaces)
+    @pytest.mark.parametrize("return_type", [Expectation, "Hermitian", "Projector"])
+    @pytest.mark.parametrize("shots", shotss)
+    def test_all_adjoint_exp(self, interface, return_type, shots):
+        """Test diff_method=adjoint works for expectation measurements for all interfaces"""
+        warn_msg = (
+            "Requested adjoint differentiation to be computed with finite shots. "
+            "Adjoint differentiation always calculated exactly."
+        )
+
+        # Hermitian doesn't work with torch at the moment
+        if interface == "torch" and return_type == "Hermitian":
+            with pytest.raises(RuntimeError):
+                with pytest.warns(UserWarning, match=warn_msg):
+                    circuit = get_qnode(interface, "adjoint", return_type, shots)
+                    x = get_variable(interface)
+                    grad = compute_gradient(x, interface, circuit, return_type)
+            return
+
+        if shots is None:
+            # test that everything runs
+            # correctness is already tested in other test files
+            circuit = get_qnode(interface, "adjoint", return_type, shots)
+            x = get_variable(interface)
+            grad = compute_gradient(x, interface, circuit, return_type)
+        else:
+            # test warning is raised when shots > 0
+            with pytest.warns(UserWarning, match=warn_msg):
+                circuit = get_qnode(interface, "adjoint", return_type, shots)
+                x = get_variable(interface)
+                grad = compute_gradient(x, interface, circuit, return_type)
+
+    @pytest.mark.parametrize("interface", diff_interfaces)
+    @pytest.mark.parametrize(
+        "return_type", [Probability, Expectation, "Hermitian", "Projector", Variance]
+    )
+    @pytest.mark.parametrize("shots", shotss)
+    def test_all_paramshift_nonstate(self, interface, return_type, shots):
+        """Test diff_method=parameter-shift works for all interfaces and
+        return_types except State and DensityMatrix"""
```

Comment:
What was the behaviour here again? If it's an incorrect result being provided to the user or a confusing error, then we should have an issue open about it in our repo.

Reply:
I think we discussed that we don't have the theory yet for using the parameter-shift rule for state gradients. Right now it raises an error which is the intended behaviour, and the error message is descriptive enough: `ValueError: Computing the gradient of circuits that return the state is not supported.`